### ISSUE TITLE
New Rule: `let`/`const`/`var`/etc. should always be required when defining a new variable.

### ISSUE THREAD

**Issue by TheAlphaReturns** [eslint/eslint#15942]
### Rule details

Make it so that a keyword (e.g. `let`) should always be used before defining a new variable

### Related ECMAScript feature

Typescript needs this.

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// bad code example, (e wasn't defined using a declaration keyword)
e = 7
e = 8

// good code example (notice the "let")
let e = 7
e = 8
```

### Why should this rule be in the core instead of a plugin?

It is required for typescript to function. Also required if you are using ESM in your node.js project.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by mdjermanovic** [comment]
Hi @TheAlphaReturns, thanks for the issue!

 

> ```js
> // bad code example, (e wasn't defined using a declaration keyword)
> e = 7
> e = 8
> 
> // good code example (notice the "let")
> let e = 7
> e = 8
> ```

I believe this is already covered by the [no-undef](https://eslint.org/docs/rules/no-undef) rule.

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVuZGVmOiBcImVycm9yXCIgKi9cblxuLy8gYmFkIGNvZGUgZXhhbXBsZSwgKGUgd2Fzbid0IGRlZmluZWQgdXNpbmcgYSBkZWNsYXJhdGlvbiBrZXl3b3JkKVxuZTEgPSA3XG5lMSA9IDhcblxuLy8gZ29vZCBjb2RlIGV4YW1wbGUgKG5vdGljZSB0aGUgXCJsZXRcIilcbmxldCBlMiA9IDdcbmUyID0gOCIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnsibm9kZSI6ZmFsc2V9fX0=)

**Comment by TheAlphaReturns** [comment]
oh, I see. Could we make it so that it would automatically add `let` when using `--fix`?

**Comment by mdjermanovic** [comment]
`--fix` should never change the behavior of the code, even if the change would be such as to avoid runtime errors, so we can't automatically add `let`.

Also, an undefined variable is often a typo in the reference, which should be corrected to match an existing variable name rather than adding a new declaration.

**Comment by nzakas** [comment]
Seems like we are already covered here, so closing.
